### PR TITLE
Added ARM coproc registers

### PR DESCRIPTION
**Pull request by mothran**
As per #652 I added some of the arm coprocessor register.  I did not map them all but this does address the two issues raised previously (#446 and #571) and allows for user to enable the VFP without starting the emu first.

### CODE REVIEW

**Comment by aquynh** [comment]
is there any existing testcase that failed before, but passes now thanks to this update?

**Comment by mothran** [comment]
No specific tests, I referenced the previous issues because the solution for those was to run a small set of instructions to enable features in the emulator.  I wanted to be able to directly access those registers instead, thus speeding up my initialization time.

For example I can enable the ARM VFP by simply doing:
```C
// Enable the CP registers for the VFP.
err = uc_reg_read(m_unicorn, UC_ARM_REG_C1_C0_2, &tmp_val);
if (err) {
	break;
}
tmp_val = tmp_val | (0xf << 20);
err = uc_reg_write(m_unicorn, UC_ARM_REG_C1_C0_2, &tmp_val);
if (err) {
	break;
}

size_t enable_vfp = 0x40000000;
err = uc_reg_write(m_unicorn, UC_ARM_REG_FPEXC, &enable_vfp);
if (err) {
	break;
}
```

I also needed write access to the ARM coprocessor that stores the TLS data for android code to run smoothly.

**Comment by aquynh** [comment]
looks good, would you put the script above into suite/regress/, as an example?

**Comment by aquynh** [comment]
ping ...

**Comment by mothran** [comment]
Updated my branch to be in sync with master and added a regression test.

**Comment by aquynh** [comment]
looks good now, but my only concern is that: by directly writing to VFP registers, you may skip some internal CPU setup that is performed by "official" initialization method, which can affect CPU operations later on. is it possible for this case?

**Comment by mothran** [comment]
Possibly but my thought here is that the VFP is only a small subset of the co-processor registers.  I added the ones I needed currently but wanted to at least leave it as a model for people to be able to add more as they go.

As for VFP setup, I would argue that from the stand point of a emulator API it is much less clean to setup some CPU features using injected instructions vs a API call.

I will do some more digging and make sure this is the appropriate setup path for the VFP and update the test to match the setup setup.

**Comment by aquynh** [comment]
Never doubt this is useful, just worry if it leaves CPU in a broken state. Let me know what you find.

**Comment by hedger** [comment]
I'd love to see this in master. Waiting for merge

File: include/unicorn/arm.h [hunk 1]
```diff
@@ -24,6 +24,9 @@ typedef enum uc_arm_reg {
 	UC_ARM_REG_FPSCR_NZCV,
 	UC_ARM_REG_FPSID,
 	UC_ARM_REG_ITSTATE,
+	UC_ARM_REG_C1_C0_2,
+	UC_ARM_REG_C13_C0_2,
```
Comment (aquynh) [inline include/unicorn/arm.h]: did you add this `UC_ARM_REG_C13_C0_2`, but never used that in `qemu/target-arm/unicorn_arm.c` ?

**Comment by mothran** [comment]
Looks like I did, would you like me to remove it? or add an implementation?  I currently have no use for that register but I can add it for posterity if needed.

**Comment by aquynh** [comment]
Yes please add it, together with a sample for it. I will merge this then.

**Comment by mothran** [comment]
Great, I will update my branch soon.

As per my comment earlier, I dug into the ARM documents and found references for enabling the VFPU:
http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.ddi0439b/BEHBJHIG.html
and FPEXC:
http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dui0068b/BCFHHBDD.html

So I think the example setup should match official method for enabling the VFPU.

**Comment by aquynh** [comment]
merged this for RC3 release, but please follow up with the sample for `UC_ARM_REG_C13_C0_2`, thanks.

**Pull request merged**